### PR TITLE
feat: [CDS-46874]: added command flags support in k8s steps

### PR DESCRIPTION
**Pull request by vikrant-harness**
### Summary

- Now the K8s Steps support the addition of command flags to apply , rolling , BG , canary steps.
- The apply step by default has the apply command selected.
- The commands for the other steps will be fetched from the API response.
- The entire change is behind feature flag :- ```NG_K8_COMMAND_FLAG```

#### Screenshots

https://user-images.githubusercontent.com/106532291/204234811-f198aaf0-9236-4c6e-971a-72b65a67e9a0.mov

image

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

**Comment by vikrant-harness** [comment]
retrigger jest

**Comment by vikrant-harness** [comment]
retrigger jest

File: src/modules/70-pipeline/components/PipelineInputSetForm/StageInputSetForm.tsx [hunk 1]
```diff
@@ -159,6 +160,45 @@ function StepFormInternal({
...
+        return (
+          <div className={cx(stepCss.formGroup, stepCss.md)}>
+            <FormInput.MultiTextInput
+              disabled={readonly}
+              name={`${path}.spec.commandFlags`}
+              multiTextInputProps={{
+                expressions,
+                allowableTypes
+              }}
+              label={`'Apply': ${getString('flag')}`}
```
Comment (rameshx) [inline src/modules/70-pipeline/components/PipelineInputSetForm/StageInputSetForm.tsx:176]: add Apply to strings

File: src/modules/70-pipeline/components/PipelineInputSetForm/StageInputSetForm.tsx [hunk 2]
```diff
@@ -159,6 +160,45 @@ function StepFormInternal({
...
+          </div>
+        )
+      }
+    }
+    return commandFlags?.map((commandFlag: CommandFlags, flagIdx: number) => {
+      if (
+        getMultiTypeFromValue(get(template, `step.spec.commandFlags[${flagIdx}].flag`)) === MultiTypeInputType.RUNTIME
+      ) {
+        return (
+          <div className={cx(stepCss.formGroup, stepCss.md)}>
```
Comment (rameshx) [inline src/modules/70-pipeline/components/PipelineInputSetForm/StageInputSetForm.tsx:187]: missing key prop?

File: src/modules/70-pipeline/components/PipelineSteps/AdvancedSteps/CommandFlagsPanel/CommandFlagsPanel.tsx [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+  })
+
+  const [commandFlagOptions, setCommandFlagOptions] = useState<SelectOption[]>([])
+  React.useEffect(() => {
+    const commandFlags = get(data, 'data', []).map((flag: string) => ({
+      label: flag,
+      value: flag
+    }))
+    setCommandFlagOptions(commandFlags as SelectOption[])
+  }, [data])
```
Comment (rameshx) [inline src/modules/70-pipeline/components/PipelineSteps/AdvancedSteps/CommandFlagsPanel/CommandFlagsPanel.tsx:61]: Remove `useEffect` and `useState` and use `useMemo` and map `data` inside it. This helps in reducing re-renders.

File: src/modules/70-pipeline/components/PipelineSteps/AdvancedSteps/CommandFlagsPanel/__tests__/CommandFlagsPanel.test.tsx [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    )
+    const addBtn = getByText(container, 'add')
+    await waitFor(() => expect(addBtn).toBeInTheDocument())
+    fireEvent.click(addBtn)
+    const defaultSelectDropdown = getByPlaceholderText('- pipeline.fieldPlaceholders.commandType -')
+    await waitFor(() => expect(defaultSelectDropdown).toBeInTheDocument())
+    fireEvent.click(addBtn)
+    const deleteBtn = getByTestId(container, 'deleteCommandFlag')
+    await waitFor(() => expect(deleteBtn).toBeInTheDocument())
+    fireEvent.click(deleteBtn)
```
Comment (rameshx) [inline src/modules/70-pipeline/components/PipelineSteps/AdvancedSteps/CommandFlagsPanel/__tests__/CommandFlagsPanel.test.tsx:108]: An assertion can be added to test if something is deleted after clicking delete

File: src/modules/70-pipeline/components/PipelineInputSetForm/StageInputSetForm.tsx (hunk 1 above)
Comment (vikrant-harness) [reply src/modules/70-pipeline/components/PipelineInputSetForm/StageInputSetForm.tsx:176]: done

File: src/modules/70-pipeline/components/PipelineInputSetForm/StageInputSetForm.tsx (hunk 2 above)
Comment (vikrant-harness) [reply src/modules/70-pipeline/components/PipelineInputSetForm/StageInputSetForm.tsx:187]: done

File: src/modules/70-pipeline/components/PipelineSteps/AdvancedSteps/CommandFlagsPanel/CommandFlagsPanel.tsx (hunk 3 above)
Comment (vikrant-harness) [reply src/modules/70-pipeline/components/PipelineSteps/AdvancedSteps/CommandFlagsPanel/CommandFlagsPanel.tsx:61]: done

File: src/modules/70-pipeline/components/PipelineSteps/AdvancedSteps/CommandFlagsPanel/__tests__/CommandFlagsPanel.test.tsx (hunk 4 above)
Comment (vikrant-harness) [reply src/modules/70-pipeline/components/PipelineSteps/AdvancedSteps/CommandFlagsPanel/__tests__/CommandFlagsPanel.test.tsx:108]: done

**Review by rameshx — APPROVED** [review]

**Review by Eric761 — DISMISSED** [review]

File: src/modules/70-pipeline/components/PipelineSteps/AdvancedSteps/AdvancedSteps.tsx [hunk 5]
```diff
@@ -200,6 +214,14 @@ export function AdvancedTabForm(props: AdvancedTabFormProps): React.ReactElement
               }
             />
           )}
+          {!hiddenPanels.includes(AdvancedPanels.CommandFlags) &&
+          stepsFactory.getStep(stepType)?.hasCommandFlagSelectionVisible ? (
+            <Accordion.Panel
+              id={AdvancedPanels.CommandFlags}
+              summary={'Command Flags'}
```
Comment (PratyushGar) [inline src/modules/70-pipeline/components/PipelineSteps/AdvancedSteps/AdvancedSteps.tsx:221]: Add summary text to strings

File: src/modules/70-pipeline/components/PipelineSteps/AdvancedSteps/AdvancedSteps.tsx [hunk 6]
```diff
@@ -69,11 +72,19 @@ export default function AdvancedSteps(props: AdvancedStepsProps, formikRef: Step
 
   const strategy = (step as any)?.strategy
 
+  const getCommandFlagInitialValue = (): CommandFlags[] => {
+    if (typeof (step as StepElementConfig)?.spec?.commandFlags === 'string') {
+      return [{ commandType: 'Apply', flag: (step as StepElementConfig)?.spec?.commandFlags }]
```
Comment (PratyushGar) [inline src/modules/70-pipeline/components/PipelineSteps/AdvancedSteps/AdvancedSteps.tsx:77]: optional assertions can be omitted as above typeof will fail for undefined

File: src/modules/70-pipeline/components/PipelineSteps/AdvancedSteps/AdvancedSteps.tsx (hunk 5 above)
Comment (vikrant-harness) [reply src/modules/70-pipeline/components/PipelineSteps/AdvancedSteps/AdvancedSteps.tsx:221]: done

File: src/modules/70-pipeline/components/PipelineSteps/AdvancedSteps/AdvancedSteps.tsx (hunk 6 above)
Comment (vikrant-harness) [reply src/modules/70-pipeline/components/PipelineSteps/AdvancedSteps/AdvancedSteps.tsx:77]: done

**Comment by vikrant-harness** [comment]
retrigger jest

**Comment by vikrant-harness** [comment]
retrigger coverage

File: src/modules/70-pipeline/components/PipelineSteps/AdvancedSteps/CommandFlagsPanel/CommandFlagsPanel.tsx [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+                            })}
+                            onClick={() => remove(index)}
+                          />
+                        )}
+                      </Layout.Horizontal>
+                    </div>
+                  </Layout.Horizontal>
+                ))}
+                {!!(get(formik, 'values.commandFlags', []).length < commandFlagOptions.length) && (
+                  <span>
```
Comment (PratyushGar) [inline src/modules/70-pipeline/components/PipelineSteps/AdvancedSteps/CommandFlagsPanel/CommandFlagsPanel.tsx:128]: Boolean assertion can be omitted as the operator returns a boolean value

**Review by rameshx — APPROVED** [review]

**Review by PratyushGar — DISMISSED** [review]

**Review by satyam-harness — APPROVED** [review]

**Review by Eric761 — DISMISSED** [review]

**Review by kapilkumawat86 — APPROVED** [review]

**Review by rameshx — DISMISSED** [review]

**Review by PratyushGar — DISMISSED** [review]

**Review by rameshx — APPROVED** [review]

**Review by YounixIO — APPROVED** [review]

**Pull request merged**